### ISSUE TITLE
[BUG] externalTrafficPolicy: Local doens't work

### ISSUE THREAD

**Issue by xolott** [Azure/AKS#3699]
**Describe the bug**
I created 3 AKS clusters using terraform with:

- Linkerd
- Cilium
- Cert Manager
- Ingress NGinx Controller

Only one of these 3 clusters gets the client's real IP. I can see them in the nginx logs.

I added the `externalTrafficPolicy: Local` setting to the ingress controller chart.

**To Reproduce**
1. Creates a new AKS cluster
2. Deploy nginx with:

```hcl

resource "helm_release" "nginx_ingress_controller" {
  name             = "ingress-nginx"
  repository       = "https://kubernetes.github.io/ingress-nginx"
  chart            = "ingress-nginx"
  namespace        = var.namespace
  create_namespace = true

  set {
    name  = "controller.admissionWebhooks.enabled"
    value = "false"
  }

  set {
    name  = "controller.podAnnotations.linkerd\\.io/inject"
    value = "enabled"
  }

  set {
    name  = "controller.service.annotations.service\\.beta\\.kubernetes\\.io/azure-load-balancer-health-probe-request-path"
    value = "/healthz"
  }

  set {
    name  = "controller.service.externalTrafficPolicy"
    value = "Local"
  }

  set {
    name  = "controller.replicaCount"
    value = var.replicas_count
  }

  set {
    name  = "controller.extraArgs.ingress-class"
    value = "nginx"
  }

  set {
    name  = "controller.ingressClassResource.name"
    value = "nginx"
  }

  set {
    name  = "controller.metrics.enabled"
    value = var.metrics_enabled
  }
}
```

**Expected behavior**
I should get the client IP

**Environment (please complete the following information):**
 - Kubernetes version: 1.25.6
 - Browser: Chrome

**Comment by abarqawi** [comment]
@xolott what is the IP for client you see in ingress controller logs ?

**Comment by xolott** [comment]
The IP I see is the Nginx controller POD IP.

**Comment by abarqawi** [comment]
@xolott i used the following and it worked fine , can you repro please ?
i will try to repro the terraform script and see

first install the ingress nginx with below command


helm install my-nginx ingress-nginx/ingress-nginx --set controller.service.externalTrafficPolicy=Local   --set controller.service.annotations."service\.beta\.kubernetes\.io/azure-load-balancer-health-probe-request-path"=/healthz 


2)install sample app 
kubectl apply -f https://raw.githubusercontent.com/abarqawi/JordanAKS/main/app.yaml


get the controller logs 
and i can see the client public IP 

kubectl logs **my-nginx-ingress-nginx-controller-8c9f48878-jw4j9** 
k8s 1.25.6

![image](https://github.com/Azure/AKS/assets/16225699/3313f032-07bc-459c-9833-36dec54bcb2b)

-----------------------
i tried also using terraform and worked fine same as your script

resource "helm_release" "nginx_ingress_controller" {
  name             = "ingress-nginx"
  repository       = "https://kubernetes.github.io/ingress-nginx"
  chart            = "ingress-nginx"
  namespace        = "default"

  set {
    name  = "controller.admissionWebhooks.enabled"
    value = "false"
  }

  set {
    name  = "controller.podAnnotations.linkerd\\.io/inject"
    value = "enabled"
  }

  set {
    name  = "controller.service.annotations.service\\.beta\\.kubernetes\\.io/azure-load-balancer-health-probe-request-path"
    value = "/healthz"
  }

  set {
    name  = "controller.service.externalTrafficPolicy"
    value = "Local"
  }

  
  set {
    name  = "controller.extraArgs.ingress-class"
    value = "nginx"
  }

  set {
    name  = "controller.ingressClassResource.name"
    value = "nginx"
  }

}
are you facing the similar?
the bottom line here what IP are you getting in ingress controller log pod , it should be the public IP of the client

**Comment by xolott** [comment]
This is the service yaml. I redacted the public IP.

<details>
<summary>Ingress NGINX controller service</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: ingress-nginx
    meta.helm.sh/release-namespace: ingress-controller
    service.beta.kubernetes.io/azure-load-balancer-health-probe-request-path: /healthz
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: ingress-nginx
    app.kubernetes.io/part-of: ingress-nginx
    app.kubernetes.io/version: 1.8.0
    helm.sh/chart: ingress-nginx-4.7.0
  name: ingress-nginx-controller
  namespace: ingress-controller
  resourceVersion: "12404"
  uid: 2d28df60-5279-48c6-aaf9-69d013926dd8
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.0.218.208
  clusterIPs:
  - 10.0.218.208
  externalTrafficPolicy: Local
  healthCheckNodePort: 31497
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - appProtocol: http
    name: http
    nodePort: 31669
    port: 80
    protocol: TCP
    targetPort: http
  - appProtocol: https
    name: https
    nodePort: 32621
    port: 443
    protocol: TCP
    targetPort: https
  selector:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/name: ingress-nginx
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 999.999.999.999
```
</details>

You can see that the settings are okay. But the IP printed in the logs is an internal IP (the ingress controller IP)

![image](https://github.com/Azure/AKS/assets/5204776/1cb16de2-57bf-43c2-b688-1f149a216b33)

This is the deployment yaml

<details>
<summary>Ingress NGINX controller deployment</summary>
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-nginx-controller
  namespace: ingress-controller
  uid: 490c77c7-ec45-450b-a220-6c36fd2e6077
  resourceVersion: '1473964'
  generation: 1
  creationTimestamp: '2023-06-01T21:48:25Z'
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: ingress-nginx
    app.kubernetes.io/part-of: ingress-nginx
    app.kubernetes.io/version: 1.8.0
    helm.sh/chart: ingress-nginx-4.7.0
  annotations:
    deployment.kubernetes.io/revision: '1'
    meta.helm.sh/release-name: ingress-nginx
    meta.helm.sh/release-namespace: ingress-controller
spec:
  replicas: 3
  selector:
    matchLabels:
      app.kubernetes.io/component: controller
      app.kubernetes.io/instance: ingress-nginx
      app.kubernetes.io/name: ingress-nginx
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: controller
        app.kubernetes.io/instance: ingress-nginx
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: ingress-nginx
        app.kubernetes.io/part-of: ingress-nginx
        app.kubernetes.io/version: 1.8.0
        helm.sh/chart: ingress-nginx-4.7.0
      annotations:
        linkerd.io/inject: enabled
    spec:
      containers:
        - name: controller
          image: >-
            registry.k8s.io/ingress-nginx/controller:v1.8.0@sha256:744ae2afd433a395eeb13dc03d3313facba92e96ad71d9feaafc85925493fee3
          args:
            - /nginx-ingress-controller
            - '--publish-service=$(POD_NAMESPACE)/ingress-nginx-controller'
            - '--election-id=ingress-nginx-leader'
            - '--controller-class=k8s.io/ingress-nginx'
            - '--ingress-class=nginx'
            - '--configmap=$(POD_NAMESPACE)/ingress-nginx-controller'
            - '--ingress-class=nginx'
          ports:
            - name: http
              containerPort: 80
              protocol: TCP
            - name: https
              containerPort: 443
              protocol: TCP
            - name: metrics
              containerPort: 10254
              protocol: TCP
          env:
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.name
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
            - name: LD_PRELOAD
              value: /usr/local/lib/libmimalloc.so
          resources:
            requests:
              cpu: 100m
              memory: 90Mi
          livenessProbe:
            httpGet:
              path: /healthz
              port: 10254
              scheme: HTTP
            initialDelaySeconds: 10
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 5
          readinessProbe:
            httpGet:
              path: /healthz
              port: 10254
              scheme: HTTP
            initialDelaySeconds: 10
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 3
          lifecycle:
            preStop:
              exec:
                command:
                  - /wait-shutdown
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
          securityContext:
            capabilities:
              add:
                - NET_BIND_SERVICE
              drop:
                - ALL
            runAsUser: 101
            allowPrivilegeEscalation: true
      restartPolicy: Always
      terminationGracePeriodSeconds: 300
      dnsPolicy: ClusterFirst
      nodeSelector:
        kubernetes.io/os: linux
      serviceAccountName: ingress-nginx
      serviceAccount: ingress-nginx
      securityContext: {}
      schedulerName: default-scheduler
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 25%
      maxSurge: 25%
  revisionHistoryLimit: 10
  progressDeadlineSeconds: 600
```
</details>

This is how I created the cluster:

<details>
<summary>Terraform plan for AKS</summary>

```hcl
data "azurerm_resource_group" "resource_group" {
  name = var.resource_group
}

resource "azurerm_user_assigned_identity" "kubelet" {
  name                = local.kubelet_identity_name
  location            = var.location
  resource_group_name = var.resource_group
}

resource "azurerm_user_assigned_identity" "aks" {
  name                = local.aks_identity_name
  location            = var.location
  resource_group_name = var.resource_group
}

resource "azurerm_role_assignment" "aks_role" {
  scope                = data.azurerm_resource_group.resource_group.id
  role_definition_name = "Managed Identity Operator"
  principal_id         = azurerm_user_assigned_identity.aks.principal_id
}

resource "azurerm_kubernetes_cluster" "aks" {
  name                = local.cluster_name
  location            = var.location
  resource_group_name = var.resource_group
  dns_prefix          = local.dns_prefix

  default_node_pool {
    name                = "default"
    node_count          = var.node_count
    vm_size             = var.vm_size
    enable_auto_scaling = var.auto_scaling_enabled
  }

  network_profile {
    network_plugin = "none"
  }


  identity {
    type = "UserAssigned"
    identity_ids = [azurerm_user_assigned_identity.aks.id]
  }

  kubelet_identity {
    client_id                 = azurerm_user_assigned_identity.kubelet.client_id
    object_id                 = azurerm_user_assigned_identity.kubelet.principal_id
    user_assigned_identity_id = azurerm_user_assigned_identity.kubelet.id
  }

  key_vault_secrets_provider {
    secret_rotation_enabled = true
  }

  oidc_issuer_enabled       = true
  workload_identity_enabled = true

  oms_agent {
    log_analytics_workspace_id = azurerm_log_analytics_workspace.default.id
  }

  tags = {
    "Environment" = var.environment
  }

  lifecycle {
    prevent_destroy = true
  }
}
```
</details>

Note that I didn't use a network plugin because then I installed cilium with hubble

<details>
<summary>Terraform plan for cilium</summary>

```hcl
resource "helm_release" "cilium" {
  name       = "cilium"
  repository = "https://helm.cilium.io/"
  chart      = "cilium"
  version    = var.cilium_version
  namespace  = "kube-system"

  set {
    name  = "aksbyocni.enabled"
    value = true
  }

  set {
    name  = "nodeinit.enabled"
    value = true
  }

  set {
    name  = "hubble.relay.enabled"
    value = true
  }

  set {
    name  = "hubble.ui.enabled"
    value = true
  }
}
```
</details>

Next I added Linkerd (cni, crds, and control plane)

<details>
<summary>Terraform plan for linkerd</summary>

```hcl
resource "helm_release" "linkerd_cni" {
  depends_on = [kubernetes_namespace.linkerd_namespace]
  name       = "linkerd-cni"
  repository = "https://helm.linkerd.io/stable"
  chart      = "linkerd2-cni"
  namespace  = "linkerd"
  wait = true
}

resource "helm_release" "linkerd_crds" {
  depends_on = [helm_release.linkerd_cni]
  name       = "linkerd-crds"
  repository = "https://helm.linkerd.io/stable"
  chart      = "linkerd-crds"
  namespace  = "linkerd"
  wait = true

  set {
    name  = "cniEnabled"
    value = true
  }
}

resource "helm_release" "linkerd" {
  depends_on = [helm_release.linkerd_crds]
  name       = "linkerd-control-plane"
  repository = "https://helm.linkerd.io/stable"
  chart      = "linkerd-control-plane"
  namespace  = "linkerd"
  wait = true

  set {
    name  = "cniEnabled"
    value = true
  }

  set {
    name  = "identity.externalCA"
    value = true
  }

  set {
    name  = "identity.issuer.scheme"
    value = "kubernetes.io/tls"
  }

  set {
    name  = "disableHeartBeat"
    value = true
  }

  set {
    name = "webhookFailurePolicy"
    value = "Fail"
  }

  set {
    name = "networkValidator.connectAddr"
    value = "0.0.0.0:4140"
  }
}
```
</details>

Could the network policies cause any issues here?

**Comment by xolott** [comment]
I solved this issue. I added one more setting to the nginx controller for the linkerd proxy to ignore the inbound ports:


```yaml
controller:
  podAnnotations:
    config.linkerd.io/skip-inbound-ports: "80,443"
```


With that in place, I started to get the source IP as expected.

Thanks @abarqawi for your time on this.

Closing this ticket

**Comment by abarqawi** [comment]
@xolott really goid catch thanks for sharing..
